### ISSUE TITLE
C++ headers are not found when cross-building for Android

### ISSUE THREAD

**Issue by theodelrieu** [conan-io/conan#2049]
Hi,

I must have missed something obvious, but I've been unable to build a sample C++ code today.

I'm using this profile:

```ini
[build_requires]
android-toolchain/r16@theo/stable

[settings]
os=Android
arch=armv8
os.api_level=27
compiler=clang
compiler.version=5.0
compiler.libcxx=libc++
build_type=Release

[options]
[scopes]
[env]
```

The compiler complains about `<iostream>` being missing...

Shouldn't the `android-toolchain` include directories be available to other recipes?
The `self.cpp_info.cflags` are present though, I do not understand why the include dirs are not seen on the command line.

**Comment by theodelrieu** [comment]
Sure, here is the full libsodium recipe which should reproduce the issue in the `test_package` step.

[android-bug.zip](https://github.com/conan-io/conan/files/1490890/android-bug.zip)

I used this command line:

>conan create theo/stable -s os=Android -s compiler=clang -s compiler.version=3.8 -s compiler.libcxx=libc++ -s arch=armv8 -s os.api_level=21 --build missing

**Comment by memsharded** [comment]
I think it might be that you need:

```cmake
target_link_libraries(example CONAN_PKG::libsodium CONAN_PKG::android-toolchain)
```

``build_requirements`` are not transitive, so depending on libsodium doesn't make available the android-toolchain by default. If you were using the old CMake approach without targets, it might work directly, but otherwise you need to define all.

You might also use ``conan_target_link_libraries(example)`` which will automatically link with all targets.

Please tell me if this makes sense (in my case, after solving this issue, it stopped with other error: 
```bash
clang38++: error: unknown argument: '-target armv8-a-linux-android'
```

**Comment by theodelrieu** [comment]
Oh yeah, don't know why I didn't use `conan_target_link_libraries`...

The next error was because of the recipe, I've patched it but now I have another issue.
The `conanbuildinfo.cmake` is adding `-stdlib=libc++` to the `CMAKE_CXX_FLAGS`, however this adds to the linker invocation `-lc++`, which does not exist in the toolchain. Removing this flag makes the linker happy.

I removed the `conan_set_libcxx` macro and everything went fine, I'm not sure if it's up to Conan to set up those flags. Meanwhile I can unset `CONAN_LIBCXX` to workaround this issue.

**Comment by memsharded** [comment]
The cons of using ``conan_target_link_libraries`` is that it will link with all defined targets, and sometimes you don't want that, you want to make explicit which targets is using each of your libraries or executables.

Regarding the automatic set of ``libc++``, the thing is that it was very problematic (the same happened to VS runtimes MD/MT, etc), because often projects don't set it, they leave it as the default decided by the build system, and then link errors difficult to debug are obtained later. So in both cases, conan will set it automatically, being possible to opt-out as you did. Maybe it makes sense to add an opt-out in the ``conan_basic_setup()`` call, like NO_OUTPUT_DIRS SKIP_RPATH arguments.

**Comment by theodelrieu** [comment]
That makes sense, however I don't know if such an option would be useful, since I just want to opt-out when using android, I'd need to have several lines in the `CMakeLists.txt` to decide which `conan_basic_setup()` to call.

There might be other use-cases though.

Would it make sense to add a few lines in the doc about this issue when dealing with Android?

**Comment by memsharded** [comment]
Yes, if only for Android, it is much more convenient to do the opt-out in the conanfile.py, that is true.
Then, I think it is not necessary to do the opt-out in CMake, unless someone really need it.

Adding to the docs make sense, but where? I guess http://docs.conan.io/en/latest/howtos/cross_building.html#cross-building-android, but as it is not really using CMake in the examples, I am not sure how to present this issue.

**Comment by theodelrieu** [comment]
What about [here](http://docs.conan.io/en/latest/howtos/cross_building.html#useful-cmake-configuration-variables)?

**Comment by memsharded** [comment]
I was writing the docs, but then I realized, why removing the ``libcxx``? So the cross-building is not linking a C++ standard library at all? Which one? If not linking a static library at all, then conan already provides a solution, which is completely remove the setting (also in that way it doesn't affect the binary ID, which is a good thing). So something like this:

```python
def configure(self):
      if self.settings.os == "Android" and self.settings.compiler == "clang":
             del self.settings.compiler.libcxx
```

**Comment by theodelrieu** [comment]
I tried to remove the `libcxx` in the lib and test_package recipes, with no success. The line `-- Conan C++ stdlib: libc++` still appears on CMake invocation.

Here is the command line:

>/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/bin/clang++ --target=aarch64-none-linux-android --gcc-toolchain=/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3 --sysroot=/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/sysroot  -funwind-tables -no-canonical-prefixes  **-stdlib=libc++** -O3 -DNDEBUG  -fPIE -pie -Wl,--gc-sections CMakeFiles/example.dir/test.cpp.o  -o bin/example  /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/aarch64-linux-android/lib/libc++_static.a /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/aarch64-linux-android/lib/libc++abi.a -static-libstdc++ /Users/theo/.conan/data/libsodium/1.0.12/theo/stable/package/9c1c63760d13d29f985f639233f0fa426bacd937/lib/libsodium.a

I emphasized the `-stdlib=libc++`. Here is the invocation (obtained with `-v`):

>"/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld" --sysroot=/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/sysroot -pie --fix-cortex-a53-843419 --eh-frame-hdr -m aarch64linux -dynamic-linker /system/bin/linker64 -o bin/example /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/sysroot/usr/lib/crtbegin_dynamic.o -L/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/lib64/clang/5.0.300080/lib/linux/aarch64 -L/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/lib/gcc/aarch64-linux-android/4.9.x -L/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/lib/../lib64 -L/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/lib -L/Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/sysroot/usr/lib --gc-sections CMakeFiles/example.dir/test.cpp.o /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/aarch64-linux-android/lib/libc++_static.a /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/aarch64-linux-android/lib/libc++abi.a /Users/theo/.conan/data/libsodium/1.0.12/theo/stable/package/9c1c63760d13d29f985f639233f0fa426bacd937/lib/libsodium.a -Bstatic **-lc++** -Bdynamic -lm -lgcc -lgcc -ldl -lc -lgcc -lgcc -ldl /Users/theo/.conan/data/android-toolchain/r16/theo/stable/package/15770c79ee9177e32ca8c27c12f5ef008fcb38c3/sysroot/usr/lib/crtend_android.o

Emphasized the `-lc++`.

I believe the `-stdlib=libc++` is causing this flag to appear. The only workaround that works for me right now is to explicitely set `CONAN_LIBCXX` to an empty string.

**Comment by memsharded** [comment]
Can you please share with me the updated version of the recipe? Would like to make sure we are sync'ed using the same. Thanks!

**Comment by theodelrieu** [comment]
Here it is, I included the updated recipes for android-ndk and android toolchain as well.

[android-bug.zip](https://github.com/conan-io/conan/files/1498990/android-bug.zip)

**Comment by memsharded** [comment]
Thanks for the info, very useful! Reproduced, indeed a bug.

**Comment by memsharded** [comment]
I am trying a fix for this in https://github.com/conan-io/conan/pull/2072

As there is a workaround, I am trying to assign it to 0.30, which will be in one or two weeks.

**Comment by memsharded** [comment]
Fixed, should be in 0.30 soon.

**Comment by JamesWeir** [comment]
I realise this issue is closed, however I came across the exact same issue, but after digging a little I found that in my case part of the problem is actually that the android make standalone toolchain copies the libc++ STL over the libstdc++ one and doesn't actually provide the correct libc++ structure, this simplifies quite a few things within Android but it makes for a problem here because CONAN_LIBCXX sets -stdlib=libc++ but it should really either be omitted or set to -stdlib=libstdc++. It would be good really to modify the cmake.py builder in Conan to understand this, though it ties it to be android version specific potentially as Google could change that behaviour.
